### PR TITLE
Add negations to all text predicates

### PR DESCRIPTION
**Pull request by WheresMyStapler**
- Add negations to all Text predicates
- Implements negation predicates for Elastic index backend
- Adds new predicate `Text.CONTAINS_PHRASE`, which exposes Elastic's `match_phrase` query
- Update Solr and Lucene to respect equivalent max edit distances as JanusGraph's `Text.CONTAINS_FUZZY.evaluateRaw(...)` and Elastic's `fuzziness:"AUTO"`

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (li-boxuan) [inline janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:265]: Any specific reason for `StringUtils.isNotBlank(condition.toString())`? I think blank space is also a valid regular expression.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (li-boxuan) [inline janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:613]: Seems this if condition always evaluates to false because you have `StringUtils.isNotBlank((String) condition)` in `isValidCondition` method.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (li-boxuan) [inline janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:689]: nitpick: Can you move this method before `private static final LevenshteinDistance ONE_LEVENSHTEIN_DISTANCE = new LevenshteinDistance(1);` because `ONE_LEVENSHTEIN_DISTANCE` is only needed by the method `isFuzzy` below.

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java
Comment (li-boxuan) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java:751]: I am just thinking if we should make JanusGraph's Text.CONTAINS_FUZZY respect Lucene's default behaviour or vice versa. Any idea?

Either way, it would be a breaking change, so we need to explain it in the changelog.md.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (li-boxuan) [inline janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:641]: Can you add the new predicates to search-predicates.md?

**Review by li-boxuan — COMMENTED** [review]
Thank you for the huge contribution! I only did a quick pass on the PR, but in general, looks good to me. Left a few comments/questions:

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (WheresMyStapler) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:265]: The existing REGEX/CONTAINS_REGEX use this test as well, so I carried it over out of consistency.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (WheresMyStapler) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:613]: Actually no, the tokenize method uses a MIN_TOKEN_LENGTH to ignore terms that are only one character long. So a short input like `a`, or one with several short terms like `a b c`, will fail out here but pass the isNotBlank check.

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java
Comment (WheresMyStapler) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java:751]: Either way works, I just wanted them to be consistent in both places. I only use JanusGraph with Elastic, which already were aligned.

I updated this so the generic testing I added worked across the board, but Ill defer to you for deciding if the change should be JG/ES or Lucene/Solr.

Ive found the default edit distance of 2 to cause some bad matching on shorter terms, which is another reason I reused JG/ES methodology.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (WheresMyStapler) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:641]: Will do.

I did have a question on text-search.md, as Ive only implemented the Elastic index backend. Should I update that page as well, or leave it out because the support across the other indexing backends is limited?

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java
Comment (li-boxuan) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java:751]: Sounds good to me. We just need to mention this in the changelog since this is a breaking change.

Comment (WheresMyStapler) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java:751]: Perfect, updated the changelog.md in the last merge. Let me know if that looks good to you, I just added a note to the end of the existing list for 0.6.0.

Comment (li-boxuan) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java:751]: > It might make sense to be explicit in the ES code instead of using AUTO, which might change implementation at some point in the future.

Is it possible to continue using AUTO, but have some test cases (if we don't have yet) to detect potential implementation change in the future? That way, we would be able to enforce the current behavior and can make a call if a future ES version changes this behavior.

Comment (WheresMyStapler) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java:751]: Yes! The generic predicate tests Ive added to IndexProviderTest already covers this, as the @MethodSource arguments provided by TextArgument include these cases.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (li-boxuan) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:641]: I think ideally, we should have it implemented across all indexing backends if possible. If you cannot do it now, maybe you could mention in text-search.md that those predicates only work for ES backend.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (li-boxuan) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:613]: Thanks, I didn't know that. Then I presume at least `!terms.isEmpty()` is always true, right?

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (li-boxuan) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:265]: Noted. Not sure why we did that in the first place, but it makes sense to follow the old behavior.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (WheresMyStapler) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:613]: I assumed that too, this is another line that I copied over from an existing predicate (CONTAINS), and I wasnt sure if there was an edge case that I wasnt aware of.

Comment (li-boxuan) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:613]: Fair enough.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (WheresMyStapler) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:641]: Updated the text-search.md file accordingly to describe the added functionality to Elastic for both text and string predicates.

As for the CLA, Ive kicked off the process to get approval/whitelist from our corporate CLA manager. Hopefully that is resolved soon.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (WheresMyStapler) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java]: Updated, and also updated all the other new predicates to use their negation's evaluateRaw as well, as they all ended up being the same code with flipped return values.

Comment (WheresMyStapler) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java]: Also reused the existing isValidCondition in each of the negations.

File: docs/changelog.md
Comment (WheresMyStapler) [reply docs/changelog.md:262]: Sure thing. Mentioned the default max-edit of 2, and added a change matrix below to show the two specific cases that differ now.

File: docs/interactions/search-predicates.md
Comment (WheresMyStapler) [reply docs/interactions/search-predicates.md]: Yes, thats much clearer. Updated all of the descriptions with similar wording.

**Review by li-boxuan — APPROVED** [review]
Thank you, LGTM 👍

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/core/attribute/TextArgument.java
Comment (farodin91) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/core/attribute/TextArgument.java:1]: 2011

**Review by farodin91 — COMMENTED** [review]
Just on question out of curiosity: Wouldn´t be better to just wrap the predicate into a `not(...)`?

**Comment by li-boxuan** [comment]
> Wouldn´t be better to just wrap the predicate into a not(...)?

@farodin91 See discussions in https://github.com/JanusGraph/janusgraph/issues/1868. Briefly speaking, not(has(key, predicate)) is not semantically equivalent to has(key, ~predicate).

**Comment by farodin91** [comment]
> > Wouldn´t be better to just wrap the predicate into a not(...)?
> 
> 
> 
> @farodin91 See discussions in https://github.com/JanusGraph/janusgraph/issues/1868. Briefly speaking, not(has(key, predicate)) is not semantically equivalent to has(key, ~predicate).

https://tinkerpop.apache.org/docs/current/reference/#a-note-on-predicates
It though of putting a not around the predicate.

**Comment by li-boxuan** [comment]
> https://tinkerpop.apache.org/docs/current/reference/#a-note-on-predicates

I see it mentions that `not(predicate)` is also a predicate. It doesn't say `not(predicate)` == negation of `predicate`, if I am understanding it correctly.

**Comment by WheresMyStapler** [comment]
> > > Wouldn´t be better to just wrap the predicate into a not(...)?
> > 
> > 
> > @farodin91 See discussions in #1868. Briefly speaking, not(has(key, predicate)) is not semantically equivalent to has(key, ~predicate).
> 
> https://tinkerpop.apache.org/docs/current/reference/#a-note-on-predicates
> It though of putting a not around the predicate.

@farodin91, wrapping the predicate in the `not()` raises an UnsupportedException, because the tinkerpop code eventually calls .negate() on the predicate (in v0.5.3):

```
gremlin> g.V().has('fullTextIndexedField', not(textContains('foo')))
java.lang.UnsupportedOperationException
Type ':help' or ':h' for help.
Display stack trace? [yN]y
java.lang.UnsupportedOperationException
	at org.janusgraph.core.attribute.Text.negate(Text.java:312)
	at org.janusgraph.core.attribute.Text$1.negate(Text.java:45)
	at org.apache.tinkerpop.gremlin.process.traversal.P.negate(P.java:98)
	at org.apache.tinkerpop.gremlin.process.traversal.P.not(P.java:257)
	at org.apache.tinkerpop.gremlin.process.traversal.P$not.callStatic(Unknown Source)
```

Since this PR implements .negate for all the text predicates, this now functions:

```
gremlin> g.V().has('fullTextIndexedField', not(textContains('foo')))
==>v[835784]
gremlin> g.V().has('fullTextIndexedField', not(textContains('foo'))).explain()
==>Traversal Explanation
============================================================================================================
Original Traversal                          [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]

ConnectiveStrategy                    [D]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
MatchPredicateStrategy                [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
EarlyLimitStrategy                    [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
IncidentToAdjacentStrategy            [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
RepeatUnrollStrategy                  [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
PathRetractionStrategy                [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
CountStrategy                         [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
FilterRankingStrategy                 [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
InlineFilterStrategy                  [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
AdjacentToIncidentStrategy            [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
LazyBarrierStrategy                   [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
JanusGraphIoRegistrationStrategy      [P]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
JanusGraphStepStrategy                [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
AdjacentVertexFilterOptimizerStrategy [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
AdjacentVertexHasIdOptimizerStrategy  [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
AdjacentVertexIsOptimizerStrategy     [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
JanusGraphLocalQueryOptimizerStrategy [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
ProfileStrategy                       [F]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
StandardVerificationStrategy          [V]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]

Final Traversal                             [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
```

But, only when using Elasticsearch, or in-memory if full scans are enabled.

**Comment by porunov** [comment]
Thank you @WheresMyStapler !
Will you be able to sign EasyCLA so that we could merge your contributions?
https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/1543822/77385607/2559/#/?version=1

**Comment by WheresMyStapler** [comment]
> Thank you @WheresMyStapler !
> Will you be able to sign EasyCLA so that we could merge your contributions?
> https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/1543822/77385607/2559/#/?version=1

Yes! Working on it now, hopefully not much longer.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/core/attribute/TextArgument.java
Comment (WheresMyStapler) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/core/attribute/TextArgument.java]: Just saw this, updated the Copyright on the new file to 2021.

**Review by porunov — APPROVED** [review]
LGTM. Thank you @WheresMyStapler !

**Comment by porunov** [comment]
@farodin91 Will you review above answers from @WheresMyStapler ?

**Comment by farodin91** [comment]
> > > > Wouldn´t be better to just wrap the predicate into a not(...)?
> 
> > > 
> 
> > > 
> 
> > > @farodin91 See discussions in #1868. Briefly speaking, not(has(key, predicate)) is not semantically equivalent to has(key, ~predicate).
> 
> > 
> 
> > https://tinkerpop.apache.org/docs/current/reference/#a-note-on-predicates
> 
> > It though of putting a not around the predicate.
> 
> 
> 
> @farodin91, wrapping the predicate in the `not()` raises an UnsupportedException, because the tinkerpop code eventually calls .negate() on the predicate (in v0.5.3):
> 
> 
> 
> ```
> 
> gremlin> g.V().has('fullTextIndexedField', not(textContains('foo')))
> 
> java.lang.UnsupportedOperationException
> 
> Type ':help' or ':h' for help.
> 
> Display stack trace? [yN]y
> 
> java.lang.UnsupportedOperationException
> 
> 	at org.janusgraph.core.attribute.Text.negate(Text.java:312)
> 
> 	at org.janusgraph.core.attribute.Text$1.negate(Text.java:45)
> 
> 	at org.apache.tinkerpop.gremlin.process.traversal.P.negate(P.java:98)
> 
> 	at org.apache.tinkerpop.gremlin.process.traversal.P.not(P.java:257)
> 
> 	at org.apache.tinkerpop.gremlin.process.traversal.P$not.callStatic(Unknown Source)
> 
> ```
> 
> 
> 
> Since this PR implements .negate for all the text predicates, this now functions:
> 
> 
> 
> ```
> 
> gremlin> g.V().has('fullTextIndexedField', not(textContains('foo')))
> 
> ==>v[835784]
> 
> gremlin> g.V().has('fullTextIndexedField', not(textContains('foo'))).explain()
> 
> ==>Traversal Explanation
> 
> ============================================================================================================
> 
> Original Traversal                          [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> 
> 
> ConnectiveStrategy                    [D]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> MatchPredicateStrategy                [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> EarlyLimitStrategy                    [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> IncidentToAdjacentStrategy            [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> RepeatUnrollStrategy                  [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> PathRetractionStrategy                [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> CountStrategy                         [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> FilterRankingStrategy                 [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> InlineFilterStrategy                  [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> AdjacentToIncidentStrategy            [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> LazyBarrierStrategy                   [O]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> JanusGraphIoRegistrationStrategy      [P]   [GraphStep(vertex,[]), HasStep([fullTextIndexedField.textNotContains(foo)])]
> 
> JanusGraphStepStrategy                [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
> 
> AdjacentVertexFilterOptimizerStrategy [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
> 
> AdjacentVertexHasIdOptimizerStrategy  [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
> 
> AdjacentVertexIsOptimizerStrategy     [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
> 
> JanusGraphLocalQueryOptimizerStrategy [P]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
> 
> ProfileStrategy                       [F]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
> 
> StandardVerificationStrategy          [V]   [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
> 
> 
> 
> Final Traversal                             [JanusGraphStep([],[fullTextIndexedField.textNotContains(foo)])]
> 
> ```
> 
> 
> 
> But, only when using Elasticsearch, or in-memory if full scans are enabled.

That's great. Thank you implementing this feature.

File: janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java
Comment (WheresMyStapler) [inline janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:398]: @porunov / @li-boxuan / @farodin91 

Sorry to bring you all back to this PR, but after using this some more locally, I ran into a case that was not covered here regarding null values / missing properties:

```
gremlin> g.V().has('missingField', textNotRegex('.*'))
==>v[86040]
gremlin> g.V(86040L).propertyMap()
==>[]
gremlin> g.V(86040L).has('missingField', textNotRegex('.*'))
gremlin>
```

Should the negation tests remove the != null tests, or just simply short circuit if null?

```
return value == null || evaluateRaw(value.toString(), (String) condition);
```

Comment (WheresMyStapler) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:398]: I went ahead and created a new PR for this change (https://github.com/JanusGraph/janusgraph/pull/2593). If you dont think `.has(X, Y)` assumes `.has(X).has(X, Y)`, feel free to close it.

Comment (li-boxuan) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:398]: > gremlin> g.V().has('missingField', textNotRegex('.*'))
==>v[86040]

I believe above is wrong. I actually suspect this will fail for TinkerPop test suite, since I don't believe this is TinkerPop-aligned. A related discussion is on https://github.com/JanusGraph/janusgraph/issues/1868

My understanding is:
`g.V().has('missingField', textNotRegex('.*')) == g.V().has('missingField', not(textRegex('.*'))) == g.V().has('missingField').has('missingField', not(textRegex('.*'))) != g.V().not(has('missingField', textRegex('.*')))`

Comment (WheresMyStapler) [reply janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java:398]: Perfect. I updated my new PR to change the negated query in ES from `mustNot(predicate)` to `must(exists(field), mustNot(predicate))`, which aligns the predicate evaluations:

```
gremlin> g.V().has('missingField', textNotRegex('.*'))
gremlin>
gremlin> g.V(86040L).propertyMap()
==>[]
gremlin> g.V(86040L).has('missingField', textNotRegex('.*'))
gremlin>
```

**Pull request merged**
